internal/events/aggregator.go

```diff
@@ -249,6 +255,46 @@ func (ag *aggregator) processPins(ctx context.Context, pins []*fftypes.Pin, stat
 	return err
 }
 
+func (ag *aggregator) checkOnchainConsistency(ctx context.Context, msg *fftypes.Message, pin *fftypes.Pin) (valid bool, err error) {
```

Comment:
The introduction of this new function, and its use in a different phase of the verification, is what I mentioned above about introducing the `signer` on the `pin`.

Previously we verified the signing identity for batches when storing them to the DB, after querying the data from IPFS - because we had the signing identity in-hand at that point. For private messages, no such code path existed.

Now we defer the check to each individual message (which retains the signer information individually) at the time we are attempting the dispatch of the pin in the aggregator. This has multiple benefits:

- We do it consistently for private and broadcast
- We do the identity lookup at the point in the sequence that the pin is being processed consistently - so that means checks on identity etc. should be validly ordered.

Note that the batch persist logic still ensures that every message in the batch, has the same signing identity.